### PR TITLE
Architecture change [Client 2.0, phase 2]

### PR DESCRIPTION
**Pull request by jellegerbrandy**
Thiis is a copy of #442 but in the current repository

### CODE REVIEW

**Comment by jellegerbrandy** [comment]
let's not forget about https://github.com/daostack/arc.js/pull/442#discussion_r417201831

**Comment by dOrgJelli** [comment]
Tests are now all passing @jellegerbrandy @orenyodfat

File: src/plugins/proposalPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    options.where.scheme = this.id
+    return Proposal.search(this.context, options, apolloQueryOptions)
+  }
+
+  protected abstract async createProposalTransaction(
+    options: TProposalCreateOptions
+  ): Promise<ITransaction>
+
+  // TODO: optional parameter 'options'?
```
Comment (dOrgJelli) [inline src/plugins/proposalPlugin.ts:80]: Remove this comment?

File: src/plugins/proposal.ts [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+  description?: string
+  descriptionHash?: string
+  title?: string
+  tags?: string[]
+  plugin?: Address
+  url?: string
+}
+
+export interface IProposalState {
+  // TODO: semantically order props
```
Comment (dOrgJelli) [inline src/plugins/proposal.ts:109]: Remove comment?

File: src/plugins/proposal.ts [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+  id: string
+  dao: IEntityRef<DAO>
+  votingMachine: Address
+  // TODO: plugin instance inside itself? or other type of plugin?
```
Comment (dOrgJelli) [inline src/plugins/proposal.ts:113]: This is the plugin of the proposal. Not a plugin with a plugin.

File: src/plugins/proposal.ts [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+          threshold
+          votingMachine
+        }
+        organizationId
+        preBoostedAt
+        proposer
+        quietEndingPeriodBeganAt
+
+        stage
+        # stakes { id }
```
Comment (dOrgJelli) [inline src/plugins/proposal.ts:213]: commenting this out?

File: documentation/demo.js [hunk 5]
```diff
@@ -13,56 +13,56 @@ async function main() {
...
+  const dao = new DAO(arc, daos[0].id)
 
   // get the DAO state (again, the "first()" object from the observable)
-  const daoState  = await dao.state().first()
+  const daoState  = await dao.fetchState()
   // the state contains information such as the address, name or nativeToken of the DAO
   console.log(`This DAO has name "${daoState.name}" and is deployed on ${daoState.address}`)
 
   // to create a proposal, we must first find the address of the Scheme to create it in
-  const schemes = await dao.schemes({ where: { name: 'ContributionReward'}}).first()
```
[the remark this answers is not in the document]
Comment (namesty) [reply documentation/demo.js:47]: Sadly it doesn't :( . Observable is not a fluid interface anymore, we have to pass the first operator into the pipe method.

File: src/utils/helpers.ts [hunk 6]
```diff
@@ -171,24 +168,26 @@ export function createGraphQlQuery(options: ICommonQueryOptions, where: string =
...
   for (const key of Object.keys(where)) {
     if (where[key] === undefined) {
       continue
     }
 
     let value = where[key]
-    if (value === null) {
-      result += `${key}: ${value}\n`
-    } else if (key === 'dao' || key === 'address') {
+    if (key === 'dao' || key === 'address') {
```
[the remark this answers is not in the document]
Comment (namesty) [reply src/utils/helpers.ts:182]: Ok, just checked all usages of `.toLowerCase` in the old codebase and I think we are fine.

File: src/plugins/utils.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+export type ProposalName = keyof typeof Proposals
+
+export type ProposalCreateOptions =
+  | IProposalCreateOptionsCRExt
+  | IProposalCreateOptionsGS
+  | IProposalCreateOptionsSR
+  | IProposalCreateOptionsComp
+  | IProposalCreateOptionsCR
+
+export type AnyProposal = Proposal<IProposalState>
```
[the remark this answers is not in the document]
Comment (namesty) [reply src/plugins/utils.ts:63]: Exactly :)

File: src/plugins/utils.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  ContributionRewardExt: ContributionRewardExtProposal,
+  SchemeRegistrar: SchemeRegistrarProposal,
+  SchemeRegistrarAdd: SchemeRegistrarProposal,
+  SchemeRegistrarEdit: SchemeRegistrarProposal,
+  SchemeRegistrarRemove: SchemeRegistrarProposal
+}
+
+export type ProposalName = keyof typeof Proposals
+
+export type ProposalCreateOptions =
```
[the remark this answers is not in the document]
Comment (namesty) [reply src/plugins/utils.ts:56]: We agreed to leave the type field for all proposal states. That exported type is to make it typesafe.

File: test/proposal-vote.spec.ts [hunk 9]
```diff
@@ -99,13 +98,15 @@ describe('Vote on a ContributionReward', () => {
     await expect(executedProposal.execute().send()).rejects.toThrow(
       // TODO: uncomment when Ethers.js supports revert reasons, see thread:
       // https://github.com/ethers-io/ethers.js/issues/446
-      /*/already executed/i*/
+      // /already executed/i
```
[the remark this answers is not in the document]
Comment (namesty) [reply test/proposal-vote.spec.ts:101]: I tried doing this, but it is still an issue with ethers (https://github.com/ethers-io/ethers.js/issues/446).

File: src/plugins/proposal.ts (hunk 4 above)
Comment (namesty) [reply src/plugins/proposal.ts:213]: `stakes` field was already commented and was not being used. Should I remove it completely?

File: documentation/demo.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (dOrgJelli) [reply documentation/demo.js:47]: well that blows :/

File: src/plugins/proposal.ts (hunk 4 above)
Comment (dOrgJelli) [reply src/plugins/proposal.ts:211]: I'm not sure, @jellegerbrandy?

**Pull request merged**